jena-arq/src/main/java/org/apache/jena/sparql/engine/main/StageGeneratorGeneric.java

```diff
@@ -65,7 +86,7 @@ protected QueryIterator execute(BasicPattern pattern, ReorderTransformation reor
                 QueryIterPeek peek = QueryIterPeek.create(input, execCxt) ;
                 // And now use this one
                 input = peek ;
-                Binding b = peek.peek() ;
+                Binding b = closeOnException(peek::peek, peek) ;
```

Comment:
It looks to me like `QueryIterPeek` should be updated to handle cancellations rather than putting execution costs into the main path.

Reply:
Mhm, an exception during `iter.peek()` could close `iter`.